feat: Allow ignoring auth key change

mprasil opened the pull request:
This adds an option to not include the authkey in state file. It might be useful for short-lived keys that are generated for each run and would otherwise lead to a "change" on every run.

The functionality is driven by `auth_key_in_state` and it defaults to True, so by default the current behavior of having the key in state is preserved.

While working on this, I decided to simplify the state handling a bit:
* Dropped the state fetch and hash comparison functionality. Behind the scenes, Ansible will do this anyways (to figure out if file needs to be updated) so we can just rely on the "changed" attribute of the state file instead.
* Set the default of `tailscale_authkey` and `tailscale_args` to empty string rather than null. This still evaluates to False in boolean context and it removes the need to use `| default()` frequently in contexts where the values might be optional.
* Removed couple `| bool` filters in conditional context where bool is implied anyways by Ansible.

mprasil commented on tasks/install.yml:
```diff
@@ -122,17 +126,15 @@
-    msg: "{{ tailscale_start.stderr | regex_replace('tskey.*\\s', 'REDACTED ') | regex_replace('\\t', '') | split('\n') }}"
+    msg: "{{ tailscale_start.stderr | default () | regex_replace('tskey.*\\s', 'REDACTED ') | regex_replace('\\t', '') | split('\n') }}"
   when:
-    - tailscale_start.changed
-    - tailscale_start.stderr | length > 0
-    - tailscale_start.failed | bool
```
Due to `changed_when: tailscale_start.rc != 0` in the tailscale_start task, task is always changed when it's failed (having non-zero result)

mprasil commented on tasks/main.yml:
```diff
@@ -14,13 +14,14 @@
   when:
-    - (not tailscale_authkey or tailscale_authkey == None) and state != "absent"
-    - not tailscale_up_skip | bool and state != "absent"
+    - not tailscale_authkey
+    - not tailscale_up_skip
+    - state != "absent"
```
This should not change the overall logic but it's hopefully more readable.

artis3n commented:
I like the idea behind these changes! Particularly, the modification with state should resolve this bug - https://github.com/artis3n/ansible-role-tailscale/issues/315 .

Maintaining the state hashing with a `template` module is smart.

My original assumption was that if the authkey changes, you need/want to re-authenticate with Tailscale, so it should always be part of the state.

> It might be useful for short-lived keys that are generated for each run and would otherwise lead to a "change" on every run.

This sounds reasonable, but can you elaborate so I can better understand your use case? What is the situation where the authkey changes very frequently, yet you don't re-authenticate the server to Tailscale? How would the playbook know when it _should_ trigger that re-authentication (by running tailscale up)? Would the user need to include `auth_key_in_state` once, then re-run and remove, triggering 2 applies to reconcile idempotence again?

I like the variable defaults as empty strings rather than null - keeps the values within their expected type, I should have done that from the beginning.

Those `| bool` were likely old ansible-lint prompts that were moved around during refactors. As long as listing passes, sounds good to me.

mprasil commented:
> What is the situation where the authkey changes very frequently, yet you don't re-authenticate the server to Tailscale?

So in my specific case I simply generate short lived pre-auth key that is only valid for ansible run. And this ansible run runs regularly and 99% of the time the node is already `tailscale up`-ed so unless some params changed, the key is simply not used at all. You don't need to re-auth, because the node was authorized in some previous run already.

There are many ways to the the short lived key, but as an example use case, one could use [Hashicorp Vault plugin for tailscale](https://github.com/davidsbond/vault-plugin-tailscale) to generate such key before running the role. (in this case the key would even be single use, so it wouldn't be useful anymore anyways)

I hope that does make sense. Essentially the idea is that instead of keeping the key around, it's only there for the minute or two the playbook runs and then it's dropped. So should it ever leak, it would be invalid in couple minutes.

mprasil commented:
@artis3n there is one test failing, but it seems to be some transient issue with `404  Not Found` when installing packages. I think you just need to re-run it again.

artis3n commented on tasks/templates/state.j2:
```diff
@@ -0,0 +1,5 @@
+{%- if auth_key_in_state -%}
+{{ (tailscale_args + '--authkey=' + tailscale_authkey) | trim | hash('sha256') }}
```
For the hash it doesn't actually matter, but it kinda bothers me anyway, if these were passed as the actual args wouldn't we need a space?

 ```suggestion
{{ (tailscale_args + ' --authkey=' + tailscale_authkey) | trim | hash('sha256') }}
```

Maybe just add a jinja comment, so it isn't interpolated into the template when rendered, calling out that this doesn't produce an actual command-line set of arguments, but serves its purpose for correctly calculating a state hash.

mprasil replied:
Well spotted, it would also bother me. 😃

I was wondering, since this doesn't really have to be the same as the command line params, perhaps we could make the state file more human friendly? Something like:

```yaml
up_args_hash: {{ tailscale_args | trim | hash('sha256') }}
{% if auth_key_in_state %}
auth_key_hash: {{ tailscale_auth_key | hash('sha256') }}
{% endif %}
```

But I'm okay to leave that for future decision and PR.

artis3n commented on tasks/install.yml:
```diff
@@ -71,27 +71,31 @@
-    - "pull"
+- name: Install | Write state file
+  ansible.builtin.template:
+    src: state.j2
```
Maybe I just never realized this - does ansible have docs explaining it auto-looks for a template file in a relative path `templates/`? Or how is ansible locating this file? From this location I'd expect a relative `../templates/state.j2`. I see the module is working, but don't really understand why

mprasil replied:
AFAIK this wasn't really documented anywhere. The closest to documentation that I've found is [this comment](https://github.com/ansible/ansible/issues/14341#issuecomment-234559431).

So for templates the lookup order is:
1. If in role:
  a. `rolename/templates/<file>`
  b. `rolename/tasks/<file>`
  c. `rolename/<file>`
2. If not in role or not found in role:
  a. `play_dir/templates/<file>`
  b. `play_dir/<file>`

Overall it's just common practice to drop templates to `templates` directory in role and refer to it by file name relative to the `templates` dir. (so normally just file name, but you can obviously nest deeper if you have many templates)

artis3n replied:
Wow that sucks. Thanks for digging up that comment

artis3n commented on tasks/install.yml:
```diff
@@ -71,27 +71,31 @@
-  ansible.builtin.include_tasks: lib/up-state.yml
-  loop:
-    - "pull"
+- name: Install | Write state file
```
Nitpick, but suggested rename of this step:

```suggestion
- name: Install | Save State
```

artis3n commented on tasks/install.yml:
```diff
@@ -71,27 +71,31 @@
+    group: "{{ ansible_user_gid }}"
+    mode: '0644'
+  register: state_file
 
 - name: Install | Bring Tailscale Up
   become: true
-  ansible.builtin.command: tailscale up {{ tailscale_up_args | trim }}
+  ansible.builtin.command: "tailscale up {{ tailscale_args | default() }} --authkey={{ tailscale_authkey }}"
```
Was removing `trim` intentional? e.g.

```
{{ tailscale_args | default() | trim }} --authkey...
```

mprasil replied:
Thanks for pointing that out, I think I missed that one when removing `| default()` because I wanted to change it to `| trim` instead, but then forgot to do it. Fixed in 89e54c2.

artis3n commented on tasks/install.yml:
```diff
@@ -122,17 +126,15 @@
     nonsensitive_stdout.skipped is not defined
     or not nonsensitive_stdout.skipped | bool
 
+- name: Install | Remove state file
```
nitpick, suggested step rename:

```suggestion
- name: Install | Revert State Upon Error
```

artis3n commented on tasks/install.yml:
```diff
@@ -122,17 +126,15 @@
     or not nonsensitive_stdout.skipped | bool
 
+- name: Install | Remove state file
+  ansible.builtin.file:
+    path: "{{ tailscale_state_folder }}/artis3n-tailscale/state"
+    state: absent
+  when:
+    - tailscale_start is failed
```
This check needs to be a little more complex. We can still simplify from `lib-up-state.yml`, but - if previous state exists, and the state changes such that `tailscale up` is run, if the command fails, we should revert to the _previous_ state hash, not clear the state file entirely.

Perhaps we use <https://docs.ansible.com/ansible/latest/collections/ansible/builtin/slurp_module.html> to reset previous content, if present. Or slurp -> compare hash -> run tailscale up -> set template afterward. The `up-state.yml` checks did have a purpose 😄

mprasil replied:
I think I have to disagree here 😅 and I suggest that this behavior represents reality better. When the `tailscale up` fails, we honestly can't really tell what the _actual_ state is as it could be anything.

Imagine this scenario:

1. User runs the role with `tailscale_args: "--shields-up"`, all works OK, state is written.
2. User runs the role without `--shields-up` and it fails in a way where connection is actually modified (so now the client does not have outgoing-only configuration applied) but still returns failing exit code. (perhaps it took slightly longer than the 60s timeout that is configured for the task) At this stage we would restore the previous state. _It would be out of sync with reality._
3. User sees failing role, reverts the change back. Now the _role parameters_ and _state_ are in sync, but those do not reflect reality. Next run ansible returns _zero_ changes, user assumes they are back with `--shields-up`, in reality node might be open to connections over VPN.

If we instead delete the state file on failure:

1. User runs the role with `tailscale_args: "--shields-up"`, all works OK, state is written.
2. User runs the role without `--shields-up` and it fails. We delete the state file.
3. User sees failing role, reverts the change back. Role will not have any prior state info, so `tailscale up` will be reapplied and `--shields-up` will be configured. State is saved and configuration, state and reality are in sync.

And in most cases with some actual failures, that leave the connection broken, the previous state is useless anyways, because the `Check if Tailscale is connected` step will detect that connection is down and will force `tailscale up` no matter what state shows.

So because of the above and because of the possibility that state file might end up being wrong, I decided to delete it if in doubt. The consequences of missing state are that we run `tailscale up` even when we didn't really have to, which is in my opinion reasonable compromise. After all, the Readme mentions, that deleting the state is safe, just not optimal:

> Deleting this directory will lead to this role re-configuring Tailscale when it is not needed, but will not otherwise break anything

I hope that makes sense to you.

mprasil replied:
To hopefully add some convincing argument.. As per [Ansible docs](https://docs.ansible.com/ansible/latest/reference_appendices/glossary.html#term-Idempotency) (emphasis mine):

> An operation is idempotent if the result of performing it once is exactly the same as the result of performing it repeatedly **without any intervening actions**.

Considering that something must intervene between the runs (either configuration change or some change on the host itself) I'd say the idempotency promise is preserved.

Having said all that, if you _really_ want to revert OG state file, we could add `backup: true` to the _Save state_ task. This will add `backup_file` field to `state_file` pointing to a backup file that was created before updating the file. (It will be something like `state.5783.2023-04-25@10:18:20~` in the same directory) So instead of removing the state file, we could just copy the backup back. So it is not a major work to implement this functionality, I just think it's incorrect in some edge cases. But if you _really_ insist, I can make the code changes for this.

artis3n replied:
I agree the action itself would be idempotent, but the role would _report_ itself as `changed`. Maybe we need to modify the `changed_when` on the `tailscale up` task?

Or...hmm, if tailscale up does nothing, I believe the exit code is always 0. So maybe we're ok 🤔

artis3n replied:
Or...no, because `changed_when` is on the return code != 0, and even if tailscale up is run, if no changes occur I believe it will report exit code 0. So maybe we are handling it already 🤔

artis3n replied:
Yeah I'm overthinking it. We're good with the current format. It's the `set state` task that would report as changed...which would be expected under this flow. The tailscale operation is unaffected.

artis3n commented:
> @artis3n there is one test failing, but it seems to be some transient issue with `404 Not Found` when installing packages. I think you just need to re-run it again.

Yeah looks like Debian 9 is having some upstream mirror issues, which I'm hoping will resolve themselves in a day or so.

For now, we can ignore this issue in the Debian 9 test, I can force merge when the PR is ready.

mprasil commented:
Hopefully I have addressed all your comments and concerns. Let me know if this PR needs more work and thank you for your feedback, really appreciate this.

artis3n reviewed (commented):
Everything looks good now! Want to think about the state behavior / idempotency comment thread we have before approving and merging to make sure I'm comfortable with that change in behavior.

artis3n commented:
Can you rebase/merge with the `main` branch?

mprasil commented:
Rebased on `main`.

artis3n commented on tasks/install.yml:
```diff
@@ -122,17 +126,15 @@
     nonsensitive_stdout.skipped is not defined
     or not nonsensitive_stdout.skipped | bool
 
+- name: Install | Revert State Upon Error
```
Given our other discussion, just rename this and we can merge!

```suggestion
- name: Install | Clear State Upon Error
```
